From 672c480146451ea815b2fa445c7e54de693b7eb7 Mon Sep 17 00:00:00 2001
From: Shaoyun Liu <Shaoyun.Liu@amd.com>
Date: Mon, 10 Apr 2017 15:33:40 -0400
Subject: [PATCH 2020/2094] adm/amdgpu: Fix PPLIB NULL function pointer when
 load amdkfd on SRIOV VF

On SRIOV VF, the PPLIB will not be enabled and hence powerplay.pp_funcs will
not be initialized which will cause kernel panic when KFD driver call back
to amdgpu for any power play related functoins.

Change-Id: Ifd1e02c2f0530f1f3b9d7cbc18a35dc13f6aaca6
Signed-off-by: Shaoyun Liu <Shaoyun.Liu@amd.com>

 Conflicts:
	drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c | 9 +++++++--
 drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c     | 2 +-
 2 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
index 5a5fd55..3b5ad75 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_amdkfd.c
@@ -344,7 +344,10 @@ void get_local_mem_info(struct kgd_dev *kgd,
 			mem_info->local_mem_size_public,
 			mem_info->local_mem_size_private);
 
-	mem_info->mem_clk_max = amdgpu_dpm_get_mclk(rdev, false) / 100;
+	if (amdgpu_sriov_vf(rdev))
+		mem_info->mem_clk_max = rdev->clock.default_mclk / 100;
+	else
+		mem_info->mem_clk_max = amdgpu_dpm_get_mclk(rdev, false) / 100;
 }
 
 uint64_t get_gpu_clock_counter(struct kgd_dev *kgd)
@@ -359,7 +362,9 @@ uint64_t get_gpu_clock_counter(struct kgd_dev *kgd)
 uint32_t get_max_engine_clock_in_mhz(struct kgd_dev *kgd)
 {
 	struct amdgpu_device *rdev = (struct amdgpu_device *)kgd;
-
+	if (amdgpu_sriov_vf(rdev))
+		return rdev->clock.default_sclk / 100;
+	else
 	/* The sclk is in quantas of 10kHz */
 		return amdgpu_dpm_get_sclk(rdev, false) / 100;
 }
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
index cc773cc..441739e 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
@@ -2356,7 +2356,7 @@ int amdgpu_vm_init(struct amdgpu_device *adev, struct amdgpu_vm *vm,
 				&adev->vm_manager.id_mgr[AMDGPU_GFXHUB];
 
 		mutex_lock(&id_mgr->lock);
-		if (adev->vm_manager.n_kfd_vms++ == 0) {
+		if ((!amdgpu_sriov_vf(adev)) && adev->vm_manager.n_kfd_vms++ == 0) {
 			/* First KFD VM: enable compute power profile */
 			if (adev->pp_enabled)
 				amdgpu_dpm_switch_power_profile(adev,
-- 
2.7.4

